### PR TITLE
Beam search algorithm implementation for TDT models

### PR DESCRIPTION
**Pull request by lilithgrigoryan**
# What does this PR do ?

PR introduces default beam search and mAES (with Ngram LM fusion) decoding implementations for TDT models.

**Collection**: ASR

# Changelog 
- [nemo/collections/asr/parts/submodules/tdt_beam_decoding.py]: Default Beam Search and mAES implementations for TDT.

# Usage
```python
python examples/asr/speech_to_text_eval.py \
  model_path="$NEMO_MODEL_PATH" \
  dataset_manifest="$DATASET_MANIFEST_PATH" \
  rnnt_decoding.strategy=maes \
  rnnt_decoding.beam.beam_size=8 \
  rnnt_decoding.beam.ngram_lm_model="$KENLM_NGRAM_MODEL_PATH" \
  rnnt_decoding.beam.ngram_lm_blank_alpha=5.0 \
  rnnt_decoding.beam.durations_beam_size=2
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+            if self.maes_prefix_alpha < 0:
+                raise ValueError("`maes_prefix_alpha` must be a positive integer.")
+
+            if self.vocab_size < beam_size + maes_expansion_beta:
+                raise ValueError(
+                    f"beam_size ({beam_size}) + expansion_beta ({maes_expansion_beta}) "
+                    f"should be smaller or equal to vocabulary size ({self.vocab_size})."
+                )
+
+            if self.maes_num_steps < 2:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:225]: Is this required? Can we decode with `maes_num_steps=1`?

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+        else:
+            self.ngram_lm = None
+
+    @typecheck()
+    def __call__(
+        self,
+        encoder_output: torch.Tensor,
+        encoded_lengths: torch.Tensor,
+        partial_hypotheses: Optional[List[Hypothesis]] = None,
+    ) -> Union[Hypothesis, NBestHypotheses]:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:256]: This type annotation is incorrect according to the real return value.
Should be something like `tuple[list[Hypothesis | NBestHypotheses],]` (tuple containing one element - a list of `Union[Hypothesis, NBestHypotheses]`)

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+        # Cache decoder results to avoid duplicate computations.
+        cache = {}
+
+        # Initialize hypothesis array with blank hypothesis.
+        start_hyp = Hypothesis(
+            score=0.0, y_sequence=[self.blank], dec_state=decoder_state, timestep=[-1], length=0, last_frame=0
+        )
+        kept_hyps = [start_hyp]
+
+        time_idx = 0
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:355]: Redundant. Please, remove

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+        Returns:
+            hypotheses: list if hypotheses without duplicates.
+        """
+        sorted_hyps = sorted(hypotheses, key=lambda x: x.score, reverse=True)
+        kept_hyps = {}
+        for hyp in sorted_hyps:
+            hyp_key = (tuple(hyp.y_sequence), int(hyp.last_frame))
+            if hyp_key in kept_hyps:
+                kept_hyp = kept_hyps[hyp_key]
+                kept_hyp.score = float(torch.logaddexp(torch.tensor(kept_hyp.score), torch.tensor(hyp.score)))
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:686]: 👍

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+
+        # Cache decoder results to avoid duplicate computations.
+        cache = {}
+
+        # Decode a batch of beam states and scores
+        beam_decoder_output, beam_state = self.decoder.batch_score_hypothesis(init_tokens, cache)
+        state = beam_state[0]
+
+        # Setup ngram LM:
+        if self.ngram_lm:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:498]: I think it we merge this block with the next `if self.ngram_lm`, it would be more readable

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+        start_hyp_kept = Hypothesis(
+            y_sequence=[self.blank],
+            score=0.0,
+            dec_state=state,
+            dec_out=[beam_decoder_output[0]],
+            timestep=[-1],
+            length=0,
+            last_frame=0,
+        )
+        if self.ngram_lm:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:512]: Init lm here (merge code above)

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 7]
```diff
@@ -0,0 +1,793 @@
...
+            # Select current iteration hypotheses
+            hyps = [x for x in kept_hyps if x.last_frame == time_idx]
+            kept_hyps = [x for x in kept_hyps if x.last_frame > time_idx]
+
+            if len(hyps) == 0:
+                continue
+
+            beam_encoder_output = encoder_outputs[:, time_idx : time_idx + 1]  # [1, 1, D]
+            # Perform prefix search to update hypothesis scores.
+            if self.zero_duration_idx != None:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:526]:
```suggestion
            if self.zero_duration_idx is not None:
```

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 8]
```diff
@@ -0,0 +1,793 @@
...
+                hyps = self.prefix_search(
+                    sorted(hyps, key=lambda x: len(x.y_sequence), reverse=True),
+                    beam_encoder_output,
+                    prefix_alpha=self.maes_prefix_alpha,
+                )  # type: List[Hypothesis]
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:531]: remove python2-style type annotations

**Review by artbataev — COMMENTED** [review]
Overall, the PR looks good! Clean code with extensive comments.

Primary suggestion: add support for `maes_num_steps == 1` (if this breaks nothing).
Also, it would be great to add tests with a tiny LM, but this is not a blocker.

Please prefer Python3-style type annotations.
Added minor comments and suggestions.

File: nemo/collections/asr/parts/submodules/tdt_beam_decoding.py [hunk 9]
```diff
@@ -0,0 +1,803 @@
...
+                    else:
+                        best_hypothesis: NBestHypotheses = NBestHypotheses(
+                            nbest_hyps
+                        )  # best_hypothesis: Hypothesis = nbest_hyps[0]
+                    hypotheses.append(best_hypothesis)
+
+        self.decoder.train(decoder_training_state)
+        self.joint.train(joint_training_state)
+
+        return (hypotheses,)
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_beam_decoding.py:322]: Here `tuple[list[Hypothesis | NBestHypotheses],]` returned, which is the correct return type

**Review by artbataev — APPROVED** [review]
Thank you so much! Cool work!
